cranelift/peepmatic/crates/runtime/src/linear.rs

```diff
@@ -0,0 +1,219 @@
+//! A linear IR for optimizations.
+//!
+//! This IR is designed such that it should be easy to combine multiple linear
+//! optimizations into a single automata.
+//!
+//! See also `src/linearize.rs` for the AST to linear IR translation pass.
+
+use crate::cc::ConditionCode;
+use crate::integer_interner::{IntegerId, IntegerInterner};
+use crate::operator::{Operator, UnquoteOperator};
+use crate::paths::{PathId, PathInterner};
+use crate::r#type::{BitWidth, Type};
+use serde::{Deserialize, Serialize};
+
+/// A set of linear optimizations.
+#[derive(Debug)]
+pub struct Optimizations {
+    /// The linear optimizations.
+    pub optimizations: Vec<Optimization>,
+
+    /// The de-duplicated paths referenced by these optimizations.
+    pub paths: PathInterner,
+
+    /// The integer literals referenced by these optimizations.
+    pub integers: IntegerInterner,
+}
+
+/// A linearized optimization.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Optimization {
+    /// The chain of increments for this optimization.
+    pub increments: Vec<Increment>,
+}
+
+/// An increment is a matching operation, the expected result from that
+/// operation to continue to the next increment, and the actions to take to
+/// build up the LHS scope and RHS instructions given that we got the expected
+/// result from this increment's matching operation. Each increment will
+/// basically become a state and a transition edge out of that state in the
+/// final automata.
```

Comment:
It would be nice to start this doc comment with a single sentence describing this type in short, followed by a blank line. Rustdoc will then show that sentence and only that sentence when visiting the documentation of the parent module.